### PR TITLE
Player Abstraction

### PR DESCRIPTION
**Pull request by aberinnj**
# Adding Player Entity

## Tasks
- [ ] Moving meat of Player functions to its own class
- [x] Generalizing leaderboard and making it support other sports

### CODE REVIEW

**Comment by aberinnj** [comment]
Still needs a bit of work. Please review though if we're going the right direction.

File: src/main/java/Entities/Player.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+		if (playerGames.containsKey(sport)) {
+			wins = playerGames.get(sport).getWins();
+			loss = playerGames.get(sport).getLoss();
+			winstreak = playerGames.get(sport).getWinstreak();
+			return new GameStats(wins, loss, winstreak);
+		}
+		return null;
+	}
+
+	public void updateGameStatsAfter_WinIn(String sport) {
```
Comment (namanshenoy) [inline src/main/java/Entities/Player.java]: What are the advantages of storing records for a player in the player model vs in the game itself?

File: src/main/java/Entities/GameStats.java [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+	}
+
+	public int getLoss(){
+		return this.loss;
+	}
+
+	public int getWinstreak() {
+		return this.winstreak;
+	}
+}
```
Comment (namanshenoy) [inline src/main/java/Entities/GameStats.java]: New Line is good practice

**Review by namanshenoy — CHANGES_REQUESTED** [review]

File: src/main/java/Entities/Player.java (hunk 1 above)
Comment (aberinnj) [reply src/main/java/Entities/Player.java]: You mean have it on all under each games? like so:
"ping pong": [player1, player2, player3],
"soccer": []
?
I was thinking:
```json
"pingpong": {
   "players": ["p1_username", "p2_username", "p3_username"], 
   "teams": [not-yet-explored], 
   "metadata": {"team-limit": 2}, 
   "houses": [not-yet-explored], 
   "brackets":["burning-sun", "lava", "blazing-fire", "smouldering-fire"]
    }
}
```
### Meanwhile in Players
```json
"ready_player_one@oracle.com": {
   "teams": {"pingpong":[], "soccer":[]},
   "house": "TootyFrootie",
   "games": [{
        "pingpong":{"win":1, "loss":24, "winstreak": 1, 
               preferences: "rely-on-honor-system"}
    }],
   "achievement": ["played-all-games", "burning-man"]
}
```
- I guess Players would be for player dashboard and stats, badges, house, personal teams?
- For viewing leaderboards and brackets, we get list of usernames from game records, and query player records for displaying player data? vice versa for the above bullet. 
- We can also store teams in game? and have player details separated, just seems easier to handle/debug; simpler to maintain?
- Separation of concerns I guess. Players keep players and their extras; games keep games and their extras. I don't know why I thought of separating em when I first saw them. Am I thinking like it's relational? 
- Functions handling all these player stuff would stay in Players. {Update win, Join Team, Join House, Change preferences, ManageUserDetailsForTypos}. Functions handling etc game details would stay in Games? {GetPlayerList, DisplayTeams, DisplayTeamPlayers, DisplayHouses, GetSportBrackets, AddNewSport}. Having it separated also has the unnecessary benefit of having smaller classes, or even smaller functions.

*Please do correct me if there's a more effective way or if I am over-engineering it again*

---

Or do you mean something else by ```storing it in the game itself```?

File: src/main/java/Entities/GameStats.java (hunk 2 above)
Comment (aberinnj) [reply src/main/java/Entities/GameStats.java]: Is it my git clrf settings? or is there really just no newline after }.

File: src/main/java/Entities/Player.java (hunk 1 above)
Comment (aberinnj) [reply src/main/java/Entities/Player.java]: That's not what we wrote though... I see, yeah scope is too small. Blind coding from the beginning. Will fix!

File: src/main/java/Entities/GameStats.java (hunk 2 above)
Comment (namanshenoy) [reply src/main/java/Entities/GameStats.java]: There's just no newline after }

**Pull request closed without merging**